File: lintcode/python/0428_pow_x_n.py
#!/usr/bin/python -t

# er fen

class Solution:
    """
    @param x {float}: the base number
    @param n {int}: the power number
    @return {float}: the result
    """
    def myPow(self, x, n):
        # write your code here
        
        if n < 0:
            x = 1/x
            n = -n
        ret = 1
        tmp = x
        
        while n > 0:
            if n%2 == 1:
                ret *= tmp
                
            tmp *= tmp
            n = n/2
            
        return ret

File: lintcode/golang/0428_pow_x_n.go
/**
 * @param x: the base number
 * @param n: the power number
 * @return: the result
 */
func myPow (x float64, n int) float64 {
    // write your code here
    if n == 0 {
        return 1
    }
    
    if n < 0 {
        x = 1/x
        n = -n
    }
    
    var ret float64
    ret = 1
    tmp := x
    
    for n > 0 {
        if n % 2 == 1 {
            ret *= tmp
        }
        
        tmp *= tmp
        n = n/2
    }
    
    return ret
}

